printk: external log buffer (CONFIG_LOGBUFFER)

darknighte opened the pull request:
This debug feature allows the kernel to use an external buffer and control
block for kernel log message. The feature is controlled by an optional
command line parameter. The existing buffer and control block can contain
existing log messages from previous boot cycles and/or the bootloader. The
command line parameter was chosen for flexibility, cross arch portability,
and the ability to dynamically enable/disable this feature. The parameter
specifies the address of a control block used to replace the default log
buffer.  Existing bootloader and kernel log messages are kept, in order,
inside the new buffer.  After a boot that preserves the buffer contents, a
bootloader can display both kernel and bootloader log entries from multiple,
previous boots. It also allows the kernel to display bootloader log entries
along with its own messages.

This feature is intended for debug purposes and has no effect unless the
command line parameter is specified.  Further, it validates the passed
control block carefully and if any checks fail, it falls back to the default
behaviour.  As such, it can be left enabled by default.
## Memory Reservation

This feature expects the bootloader to reserve/preserve the shared buffer
memory. This reservation needs to prevent the kernel from overwriting the
external log control block and log entries. In my testing, I've used the
'fdt' commands in uboot to dynamically inject reserved memory regions via
the DT to the kernel.

Based on the initial work of Wolfgang Denk and Igor Lisitsin [1].
Also based on work by: Alexander Streit Alexander.Streit@continental-corporation.com

[1] http://git.denx.de/?p=linux-2.6-denx.git;a=commitdiff;h=212f61c7fd3b952a81d1459dd32a86a32ddfd4ce

Signed-off-by: Sean Hudson sean_hudson@mentor.com
Reviewed-by: Wade Farnsworth wade_farnsworth@mentor.com
Reviewed-by: Joe MacDonald joe_macdonald@mentor.com

wfarnsworth commented on include/linux/logbuff.h:
```diff
@@ -0,0 +1,38 @@
+static void inline setup_ext_logbuff(void) { }
+#endif /* CONFIG_LOGBUFFER */
+
+#endif /* _LOGBUFF_H_ */
```
I'm wondering if there is an existing header file that this could be folded into, rather than creating a new one.

darknighte replied:
Actually, there is a printk.h, but in the process of discovering that, I found that there is already a setup_log_buf() being called in main.c, so I dropped this file entirely and added the call to setu_ext_logbuff() inside that function, with compile guards.

wfarnsworth commented on init/Kconfig:
```diff
@@ -1493,6 +1493,18 @@ config PRINTK
+	depends on PRINTK
+	help
+	  This option enables support for an alternative, "external"
+	  printk log buffer. When enabled, an architecture- or machine-
```
Is it still architecture/machine specific, if you've moved the critical sections out of arch/arm?

darknighte replied:
Good catch. I've corrected that now.

wfarnsworth commented on kernel/printk/printk.c:
```diff
@@ -331,6 +332,9 @@ enum log_flags {
 
 struct printk_log {
+#ifdef CONFIG_LOGBUFFER
+	u32 magic;
```
Might want to add a comment.  It looks like the other fields in the printk_log are commented.

jjmcdn replied:
I was also thinking it might make sense to decorate this name (eg. `lb_magic` or something, rather than just `magic`).

darknighte replied:
Done.

wfarnsworth commented on kernel/printk/printk.c:
```diff
@@ -3319,4 +3357,211 @@ void show_regs_print_info(const char *log_lvl)
+			log_cb->log_buf, log_cb->log_length, new_cb->log_buf,
+			new_cb->log_length );
+
+	// Check if log sizes match
```
Should use a C-style comment here and elsewhere in this file

darknighte replied:
Fixed.

jjmcdn commented on kernel/printk/printk.c:
```diff
@@ -353,28 +357,53 @@ DEFINE_RAW_SPINLOCK(logbuf_lock);
+
+        /* Pointer to log buffer space and length of space */
+        char *log_buf;
+        u32 log_length;
```
This looks to be a replacement for `log_buf_len` throughout the remainder of the code.  If that's the case, why is it being renamed?  Historical accident?  Does it actually behave differently or indicate something different?

darknighte replied:
Great suggestion.  Done.

jjmcdn commented on kernel/printk/printk.c:
```diff
@@ -3319,4 +3357,211 @@ void show_regs_print_info(const char *log_lvl)
+		new_cb->log_version == LOGBUFF_LOG_VERSION &&
+		new_cb->magic == LOGBUFF_MAGIC &&
+		new_cb->log_cb_pad_length ==
+		LOGBUFF_CB_PAD_LENGTH &&
```
Purely stylistic, but any reason to break this and the one below on line 3401 over two lines?

darknighte replied:
Lindent did that.

jjmcdn commented on kernel/printk/printk.c:
```diff
@@ -3319,4 +3357,211 @@ void show_regs_print_info(const char *log_lvl)
+
+static bool is_valid_external_cb ( const log_cb_t const * new_cb )
+{
+	// Validate the new control buffer
```
This is C++ you savage.  :-P

darknighte replied:
Duly noted and fixed.

jjmcdn commented on kernel/printk/printk.c:
```diff
@@ -3319,4 +3357,211 @@ void show_regs_print_info(const char *log_lvl)
+		new_cb->log_phy_addr != 0);
+}
+
+// TODO: Release memory replaced by external buffer
```
It's probably not the end of the world, you won't be the only one leaving a TODO in a kernel file, but if there's some way to phase this that's sounding less like a leak, that might be good.

darknighte replied:
Removed.

jjmcdn commented on kernel/printk/printk.c:
```diff
@@ -3319,4 +3357,211 @@ void show_regs_print_info(const char *log_lvl)
+
+	/* Check new CB pointer. Ignore the external buffer if NULL. */
+	if ( ! new_cb )
+	{
```
This is the first place I noticed it, the convention I've mostly seen throughout the kernel is to cuddle braces with the conditional.  I don't like it, but it's not my kernel.  :-)

darknighte replied:
Ran Lindent to fix.

jjmcdn commented on kernel/printk/printk.c:
```diff
@@ -3319,4 +3357,211 @@ void show_regs_print_info(const char *log_lvl)
+			new_cb->clear_idx = offset_idx;
+
+		// Update next offset and sequence
+		// Note: avoid incrementing too far by excluding the last pass
```
It's not clear to me at a first glance what this note is intending to communicate.  Is there a potential bug here?  Is it explaining the logic behind the conditional that follows?

darknighte replied:
After further consideration, the code stands alone, so I removed the comment.

Commit:
printk: collect offsets into replaceable structure

Currently, printk relies on several indices that are declared as static
global variables in printk.c.  This patch collects those into a single
structure referenced by a pointer.  This allows easier replacement of
these indices and pinning to a specific locatino.

Signed-off-by: Sean Hudson <sean_hudson@mentor.com>

Commit:
printk: external log buffer (CONFIG_LOGBUFFER)

This debug feature provides a convenient way to collect log entries across
multiple, warmboot cycles and to share those entries with a boot loader.
It allows the kernel to use an external buffer for kernel log messages and
is controlled by an optional command line parameter. The buffer can contain
existing log messages from previous boot cycles and/or the bootloader. The
command line parameter was chosen for flexibility, cross arch portability,
and the ability to dynamically enable/disable this feature. The parameter
specifies the address of a control block used to replace the default log
buffer.  Existing bootloader and kernel log messages are kept, in order,
inside the new buffer.  After a boot that preserves the buffer contents, a
bootloader can display both kernel and bootloader log entries from
multiple, previous boots. It also allows the kernel to display bootloader
log entries along with its own messages.

This feature is intended for debug purposes and has no effect unless the
command line parameter is specified.  Further, it validates the passed
control block carefully and if any checks fail, it falls back to the
default behaviour.  As such, it can be left enabled by default.

Memory Reservation
------------------
This feature expects the bootloader to reserve/preserve the shared buffer
memory. This reservation needs to prevent the kernel from overwriting the
external log control block and log entries. In my testing, I've used the
'fdt' commands in uboot to dynamically inject reserved memory regions via
the DT to the kernel.

Based on the initial work of Wolfgang Denk and Igor Lisitsin [1].
Also based on work by: Alexander Streit <Alexander.Streit@continental-corporation.com>

[1] http://git.denx.de/?p=linux-2.6-denx.git;a=commitdiff;h=212f61c7fd3b952a81d1459dd32a86a32ddfd4ce

Signed-off-by: Sean Hudson <sean_hudson@mentor.com>
